### ISSUE TITLE
Publish manifest for winget

### ISSUE THREAD

**Issue by sgade** [nickbabcock/OhmGraphite#438]
I'd like to use [winget](https://github.com/microsoft/winget-cli) to install and update OhmGraphite.

To make OhmGraphite available, a manifest needs to be published in the corresponding [repository](https://github.com/microsoft/winget-pkgs). It's a YAML file format that can easily be bootstrapped using this tool: https://github.com/microsoft/winget-create.
I just did this locally and was able to install OhmGraphite!

A few points why I raise this issue here, though, instead of just creating a PR directly:
1. I don't what to do this without your consent, since it is your application.
2. I am not 100% sure of how the internals of both tools work together exactly. I suppose it needs to _not_ override the configuration file on upgrades, for example. I would also like to see an option to install a Windows service, but afaik that is not a feature of winget currently.
3. Since the manifests are separate by version, I think it would be easiest to automate the generation and opening of PRs using GitHub Actions.

Let me know what you think and if I can be of any help.

**Comment by nickbabcock** [comment]
1. Seems like a great idea, how can I help?
2. The configuration file (`OhmGraphite.exe.config`) is meant for initial installs. There aren't breaking changes introduced to the configuration, so those updating can safely ignore it
3. I'm not quite sure what this entails. Are you saying that a release of OhmGraphite should trigger a pull request against the winget-pkgs repo? If so, sounds good.

**Comment by sgade** [comment]
1. I think the best way to go about it would be that you use the `winget-create` utility to create the manifest and ensure all properties are set correctly. I used the "portable" preset but I am not sure if that is 100% correct.
2. I haven't tested it with my local install but I can do so later and report back on how winget handles unzipping with the config already existing.
3. Yes, afaik [PRs](https://github.com/microsoft/winget-pkgs/pulls) are used to create new versions. I'm new to the ecosystem myself, so I'm not sure if automating that is feasible and allowed but it would be nice to upgrade to newer versions via the package manager instead of doing it manually (which honestly I never did on my old machine).

There are a few packages which seem to only have manifests for major versions and there are some with a manifest for each minor. I guess its up to you if you want to also create a PR on every patch version as well, and how you would handle that with your versioning scheme.

**Comment by sgade** [comment]
What my experimentation yielded are the manifest files (in the structure `n\NickBabcock\OhmGraphite\0.32.0\`) that I'll attach to this post.
[NickBabcock.OhmGraphite.installer.yaml](https://github.com/nickbabcock/OhmGraphite/files/14184044/NickBabcock.OhmGraphite.installer.yaml.txt)
[NickBabcock.OhmGraphite.locale.en-US.yaml](https://github.com/nickbabcock/OhmGraphite/files/14184045/NickBabcock.OhmGraphite.locale.en-US.yaml.txt)
[NickBabcock.OhmGraphite.yaml](https://github.com/nickbabcock/OhmGraphite/files/14184046/NickBabcock.OhmGraphite.yaml.txt)

The program is installed into `~\AppData\Local\Microsoft\WinGet\Packages\NickBabcock.OhmGraphite__DefaultSource`. I think that the `__DefaultSource` suffix will be dropped if installed from the official packages mirror.

I tested out creating a fake 0.33.0 version and running `winget upgrade`. This gives the following warning, I suppose because the configuration file has been changed:
```
Extracting archive...
Successfully extracted archive
Starting package install...
Unable to remove Portable package as it has been modified; to override this check use --force
```

To be honest, I find it a bit hard to see which options we could pass to ignore that file on upgrades. Once we figure out the correct manifest defintions, a GitHub bot using GitHub Actions could trigger a new PR whenever you push a tag/create a release with all the required information.

**Comment by nickbabcock** [comment]
Thank you for looking into it. I was hoping this could be an easy win, but it looks like this is more complicated than anticipated.

Seems like I'll need to set aside some time to dive into the winget ecosystem.

**Comment by sgade** [comment]
It is a bit opaque, to be honest.

Do you have any recommendations on how to keep the app updates otherwise? Package managers are very nice because they make it very painless to do.

**Comment by nickbabcock** [comment]
Yeah publishing to winget should really be integrated into OhmGraphite directly so that when I make a release, the package gets automatically published! I could have swore I saw a tool recently that was pretty slick to automate this, but I seem to have forgotten it.

Otherwise I'll want to look into something like `vedantmgoyal9/winget-releaser`

**Comment by sgade** [comment]
Hej again, I took the time to research some existing GitHub Actions and came across [WinGet Releaser](https://github.com/marketplace/actions/winget-releaser) as well.

I'd like to support, however, it requires the setup of a classic PAT and also the clone of the winget packages repository under the same account where the publishing package is hosted.
So I think this needs to be a setup done by you.

I'd appreciate if you could find the time to set it up as required. Just let me know if I can be of assistance.

**Comment by nickbabcock** [comment]
Yep! Have we hammered out the some of the nuances?

- Should the OhmGraphite release avoid shipping example configuration to solve "Unable to remove Portable package as it has been modified" errors? If so what do release artifacts look like and how do we make it easier to discover configuration?
- winget doesn't manage the service, so one needs to still manually stop, upgrade, and then restart OhmGraphite. Is that expected? Is this how other services in winget are managed?

**Comment by sgade** [comment]
Good points. I refreshed my memory and here is what I could gather:

The current v0.38.0 release has the following artifacts:
- `NLog.config`
- `OhmGraphite.exe`
- `OhmGraphite.exe.config`

Renaming the `OhmGraphite.exe.config` to `OhmGraphite.exe.config.sample` would keep the example configuration discoverable while avoiding the modification error from winget.

The documentation could note that the user has to first copy the file (or removing the `sample` extension) and then edit it to match their setup which is what they're currently doing anyway. I don't think this would break anything.

---

winget [checks the installer/zip hashes](https://github.com/microsoft/winget-cli/discussions/2940#discussioncomment-5015752), so I understand that is why there is no direct script support currently. I didn't find anything specific on the management of Windows services and no best practices for such a case, so I see two paths OhmGraphite could take:

1. Use winget as the distribution method only, and educate the user that they need to manage their service manually.
2. Create an installer that takes care of the service during installation/update.

I currently don't have access to my Windows machine for a few weeks and cannot test the implications of cases such as installing a new version and uninstalling the old one while its service is still running. It might be required to have `uninstall` and `install` run to avoid problems.
If possible (although not optimal), path 1 would be the easiest to implement. Also, I am not familiar with the effort required to create a modern Windows installer, and recognize that it changes the portable nature of OhmGrpahite's current distribution.

**Comment by nickbabcock** [comment]
My assumption is that packaging `.sample` files would trip the same winget modification flag when the user renames the sample (or deletes it as it is no longer needed). The only way I can think around it is to have just the exe packaged and expose an `init` command that outputs configuration files for the user to modify. Seems a bit roundabout but I think winget would be happy with that approach 🤔

**Comment by sgade** [comment]
That seems like a good solution. The documentation would point to the command to get users started, and the sample config could also be available directly on GitHub.

I believe that way users have both ways to discover how to get started.
